components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java

```diff
@@ -316,9 +320,10 @@ public void processResponse(HttpServletRequest request) throws SAMLSSOException
         XMLObject samlObject = unmarshall(decodedResponse);
         if (samlObject instanceof LogoutResponse) {
             //This is a SAML response for a single logout request from the SP
+            // TODO need to change the API of this method to prevent unmarshalling twice.
             doSLO(request);
         } else {
-            processSSOResponse(request);
+            processSSOResponse(request, (Response) samlObject);
         }
     }
```

Comment:
I think it's better if we can add an else if section and check if the samlObject is an instance of Response rather than directly casting. Unless log an error saying that it's an unknown XMLObject and does not know how to process that. That will audit if any malicious injection is being tried out.